### DISCUSSION TITLE
Help needed: Control Display-Brightness via Illumination-Sensor in Berry

### DISCUSSION THREAD

**Discussion by rug1024** [arendst/Tasmota#21261] [General]
Hi!

First of all: I am not a programmer, but trying to learn and understand. So you may laugh at me or ignore this posting,
but I have not been able to find help elsewhere. Already looked into the Berry Cookbook, studied [other scripts](http://sfromis.strangled.net/tasmota/berry/github-repositories) , etc. , even asked AI (CoPilot), but those results are not working. By the way: CoPilot can create complex working Tasmota-Rules, it just fails with Berry.

My Hardware: ESP32, MAX7219 LED Matrix Display, Brightness sensor BH1750.
What I am trying to do (in Berry-Script): set the brightness (DisplayDimmer) of the LED-Display according to the surrounding brightness (4 steps), executed every 30 seconds.
Working Berry-Commands I have found out so far: **tasmota.read_sensors("BH1750")** (output needs formatting) and **tasmota.cmd("displaydimmer 10")**

This might not be complicated, but I am currently unable to accomplish this in berry due to lack of programming knowledge.
I previously did this via ioBroker (Blockly-Script), worked fine but caused a lot of MQTT-Traffic.
Then I moved the task to the Tasmota-Device itself (using Tasmota-Rules). This is currently working fine.

```
rule1
on Time#Initialized do rule2 1 endon
on rules#timer=1 do rule2 1 endon

rule2 
ON BH1750#Illuminance<100 DO DisplayDimmer 10 ENDON 
ON BH1750#Illuminance>=100 AND BH1750#Illuminance<500 DO DisplayDimmer 30 ENDON
ON BH1750#Illuminance>=500 DO DisplayDimmer 90 ENDON
ON BH1750#Illuminance DO backlog ruletimer1 30; rule2 0 ENDON
```

But now, the next step is to port this into Berry, mainly for learning purposes and to have a base for
creating other tasks. Really looking forward to your hints. Thanks in advance! Regards...

**Comment by s-hadinger** [comment]
I don't have such sensor but here is something that should work. Try it in the Berry console and if ok, create `autoexec.be` with this content

```berry
def check_illuminance()
  import json
  var payload = json.load(tasmota.read_sensors())
  var illuminance = payload.find("BH1750",{}).find("Illuminance")
  if illuminance != nil
    var dimmer
    if   illuminance < 100      dimmer = 10
    elif illuminance < 500      dimmer = 30
    elif illuminance < 700      dimmer = 70
    else                        dimmer = 100
    end
    tasmota.cmd(f"displaydimmer {dimmer}", true)    # remove `true` argument if you want to see the commands in logs
  end
end

tasmota.add_cron("*/30 * * * * *", check_illuminance)   # use `*/10` for every 10 seconds
```

**Reply by rug1024** [reply]
Working! Really nice!
Appreciate that you took the time for this!

**Comment by rug1024** [comment]
One last wish to get this perfect... ;-)

I use those Matrix Displays around my house as clocks as they are highly visible throughout the room, but also they show when main door is opened. Really really useful then you have little kids!

![grafik](https://github.com/arendst/Tasmota/assets/96115731/370af068-ece6-49f7-8a9c-ab1c9d70e6c1) ![grafik](https://github.com/arendst/Tasmota/assets/96115731/f77b8f87-7710-4c2f-81ec-40c540ea7361)


Can the script be extended so that when button "TÜR" is activated (Power0=true) (which is set externally via MQTT from ioBroker), the dimming procedure should be paused and the following commands be executed:

`tasmota.cmd("displaydimmer 100")
tasmota.cmd("displaymode 1")
tasmota.cmd("displaytext TÜR!")`

When Power0 returns to "false" this command shall be executed to return to clock display, the dimming scheme via light sensor shall be reactivated:

`tasmota.cmd("displayclock 2")`

Thank you!

**Reply by sfromis** [reply]
To stop dimming when first Power channel is on, you can expand `check_illuminance()` with this at the start:
`if tasmota.get_power(0) return end`

And then pick up power state changes like this:
```
tasmota.add_rule('Power1#state==0',
  def ()
    tasmota.cmd("Backlog displaydimmer 100; displaymode 1; displaytext TÜR!")
  end)
tasmota.add_rule('Power1#state==1',
  def ()
    tasmota.cmd("displayclock 2")
    check_illuminance()
  end)
```

**Reply by rug1024** [reply]
Thank you very much. It's working.
